src/main/java/org/jabref/gui/JabRefFrame.java

```diff
@@ -1200,6 +1202,29 @@ private boolean confirmClose(LibraryTab libraryTab) {
             // Save was cancelled or an error occurred.
             return false;
         }
+        if (response.isPresent() && response.get().equals(discardChanges)) {
+            //filename = filename + "--discarded.bak";
+
+            Path directory = Path.of(filename);
+            try {
+                SaveDatabaseAction saveAction = new SaveDatabaseAction(libraryTab, prefs, Globals.entryTypesManager);
+                String backUpPath = BackupFileUtil.getPathOfLatestExisingBackupFile(directory, BackupFileType.BACKUP).map(Path::toString).
+                        orElse(Localization.lang("File not found"));
+                backUpPath = backUpPath.replace(".bak","--discarded.bak");
+
+                if (saveAction.saveAs(Path.of(backUpPath))) {
+                    BackupManager.setDiscardedFileExists(true);
+                    return true;
+                }
```

Comment:
I think, you misunderstood the user interaction of JabRef.

1. User starts JabRef
2. User opens a file `demo.bib`
3. User changes  file `demo.bib` for 10 minutes (and does not save)
5. User closes file.
6. JabRef diplays "File changed dialog"
7. User selects "Discard changes"
8. User opens the file `demo.bib` again
9. JabRef should NOT prompt for restoring backup

In the code at hand, sets the flag after step 7, but not at step 8, where it is needed.